source/uwp/Renderer/lib/AdaptiveColumnRenderer.cpp

```diff
@@ -3,168 +3,135 @@
-                                           _COM_Outptr_ IUIElement** ColumnControl) noexcept
-    try
+    rtxaml::UIElement AdaptiveColumnRenderer::Render(rtom::IAdaptiveCardElement const& cardElement,
+                                                     rtrender::AdaptiveRenderContext const& renderContext,
+                                                     rtrender::AdaptiveRenderArgs const& renderArgs)
     {
-        ComPtr<IAdaptiveCardElement> cardElement(adaptiveCardElement);
-        ComPtr<IAdaptiveColumn> adaptiveColumn;
-        RETURN_IF_FAILED(cardElement.As(&adaptiveColumn));
+        try
+        {
+            auto adaptiveColumn = cardElement.as<rtom::AdaptiveColumn>();
 
-        ComPtr<IBorder> columnBorder =
-            XamlHelpers::CreateABIClass<IBorder>(HStringReference(RuntimeClass_Windows_UI_Xaml_Controls_Border));
+            rtxaml::Controls::Border columnBorder{};
 
-        ComPtr<WholeItemsPanel> columnPanel;
-        RETURN_IF_FAILED(MakeAndInitialize<WholeItemsPanel>(&columnPanel));
+            auto columnPanel = winrt::make<rtrender::implementation::WholeItemsPanel>();
 
-        ComPtr<IUIElement> columnPanelAsUIElement;
-        RETURN_IF_FAILED(columnPanel.As(&columnPanelAsUIElement));
+            columnBorder.Child(columnPanel);
 
-        RETURN_IF_FAILED(columnBorder->put_Child(columnPanelAsUIElement.Get()));
+            auto previousContextRtl = renderContext.Rtl();
+            auto currentRtl = previousContextRtl;
+            auto containerRtl = adaptiveColumn.Rtl();
 
-        ComPtr<IFrameworkElement> columnPanelAsFrameworkElement;
-        RETURN_IF_FAILED(columnPanel.As(&columnPanelAsFrameworkElement));
+            bool updatedRtl = false;
 
-        ComPtr<IReference<bool>> previousContextRtl;
-        RETURN_IF_FAILED(renderContext->get_Rtl(&previousContextRtl));
-        ComPtr<IReference<bool>> currentRtl = previousContextRtl;
+            if (containerRtl != nullptr)
+            {
+                currentRtl = containerRtl;
+                renderContext.Rtl(currentRtl);
+                updatedRtl = true;
+            }
 
-        ComPtr<IReference<bool>> containerRtl;
-        RETURN_IF_FAILED(adaptiveColumn->get_Rtl(&containerRtl));
+            if (currentRtl != nullptr)
+            {
+                columnPanel.FlowDirection(currentRtl.Value() ? rtxaml::FlowDirection::RightToLeft : rtxaml::FlowDirection::LeftToRight);
+            }
 
-        bool updatedRtl = false;
-        if (containerRtl != nullptr)
-        {
-            currentRtl = containerRtl;
-            RETURN_IF_FAILED(renderContext->put_Rtl(currentRtl.Get()));
-            updatedRtl = true;
-        }
+            rtom::ContainerStyle containerStyle{};
 
-        if (currentRtl)
-        {
-            boolean rtlValue;
-            RETURN_IF_FAILED(currentRtl->get_Value(&rtlValue));
+            // TODO: do I even need this if (const auto..) ?
```

Comment:
No you don't. `AdaptiveColumn` is declared in metadata as implementing `IAdaptiveContainerBase` so C++/WinRT will do the automatic QI for you. The only case you would want to do this manually is if you are afraid that `AdaptiveColumn` doesn't actually implement `IAdaptiveContainerBase` (e.g., running downlevel on a version that doesn't implement it), but that's not the case here since we *are* the implementation.